File: src/main/scala/com/exasol/cloudetl/emitter/FilesDataEmitter.scala
package com.exasol.cloudetl.emitter

import java.io.IOException
import java.util.List
import java.util.TimeZone
import java.util.function.Consumer

import scala.util.Using

import com.exasol.ExaIterator
import com.exasol.cloudetl.bucket.Bucket
import com.exasol.cloudetl.parquet.ParquetValueConverter
import com.exasol.cloudetl.source.Source
import com.exasol.cloudetl.storage.FileFormat
import com.exasol.cloudetl.storage.StorageProperties
import com.exasol.cloudetl.transform.DefaultTransformation
import com.exasol.common.data.{Row => RegularRow}
import com.exasol.errorreporting.ExaError
import com.exasol.parquetio.data.ChunkInterval
import com.exasol.parquetio.data.Row
import com.exasol.parquetio.reader.RowParquetChunkReader
import com.exasol.parquetio.reader.RowParquetReader

import com.typesafe.scalalogging.LazyLogging
import org.apache.hadoop.fs.Path
import org.apache.parquet.hadoop.util.HadoopInputFile
import org.apache.parquet.io.InputFile

/**
 * A class that emits data read from storage into an Exasol database.
 *
 * @param properties a storage properties to create configurations
 * @param files a list of files with the chunks to read
 */
final case class FilesDataEmitter(properties: StorageProperties, files: Map[String, List[ChunkInterval]])
    extends Emitter
    with LazyLogging {

  private[this] val bucket = Bucket(properties)
  private[this] val fileFormat = properties.getFileFormat()
  private[this] val defaultTransformation = new DefaultTransformation(properties)

  override def emit(context: ExaIterator): Unit = {
    checkIfTimezoneUTC()
    emitFileFormatData(context)
  }

  private[this] def checkIfTimezoneUTC(): Unit =
    if (properties.isEnabled("TIMEZONE_UTC")) {
      TimeZone.setDefault(TimeZone.getTimeZone("UTC"))
    }

  private[this] def emitFileFormatData(context: ExaIterator): Unit =
    fileFormat match {
      case FileFormat.PARQUET => emitParquetData(context)
      case FileFormat.DELTA   => emitParquetData(context)
      case _                  => emitRegularData(context)
    }

  private[this] def emitRegularData(context: ExaIterator): Unit = {
    var totalRowCount = 0
    files.foreach { case (filename, _) =>
      Using(Source(fileFormat, new Path(filename), bucket.getConfiguration(), bucket.fileSystem)) { source =>
        var rowCount = 0
        source.stream().foreach { row =>
          val values = defaultTransformation.transform(transformRegularRowValues(row))
          context.emit(values: _*)
          rowCount += 1
        }
        totalRowCount += rowCount
        logger.info(s"Imported file $filename with $rowCount rows ($totalRowCount until now)")
      }
    }
    logger.info(s"Imported ${files.size} files with $totalRowCount rows in total")
  }

  private[this] def transformRegularRowValues(row: RegularRow): Array[Object] =
    row.getValues().map(_.asInstanceOf[Object]).toArray

  private[this] def emitParquetData(context: ExaIterator): Unit = {
    var totalRowCount = 0
    files.foreach { case (filename, intervals) =>
      val inputFile = getInputFile(filename)
      val converter = ParquetValueConverter(RowParquetReader.getSchema(inputFile))
      val source = new RowParquetChunkReader(inputFile, intervals)
      var rowCount = 0
      source.read(new Consumer[Row] {
        override def accept(row: Row): Unit = {
          val values = defaultTransformation.transform(converter.convert(row))
          context.emit(values: _*)
          rowCount += 1
        }
      })
      totalRowCount += rowCount
      logger.info(
        s"Imported file $inputFile with $rowCount rows and ${intervals.size()} intervals ($totalRowCount rows until now)"
      )
    }
    logger.info(s"Imported ${files.size} files with $totalRowCount rows in total")
  }

  private[this] def getInputFile(filename: String): InputFile =
    try {
      HadoopInputFile.fromPath(new Path(filename), bucket.getConfiguration())
    } catch {
      case exception: IOException =>
        throw new IllegalStateException(
          ExaError
            .messageBuilder("E-CSE-27")
            .message("Failed to create an input file from {{FILE}}.", filename)
            .mitigation("Please make sure that the file is not corrupted.")
            .toString(),
          exception
        )
    }

}

File: src/main/java/com/exasol/cloudetl/emitter/FilesDataEmitter.java
package com.exasol.cloudetl.emitter;

import java.io.IOException;
import java.util.*;

import org.apache.hadoop.fs.Path;
import org.apache.parquet.hadoop.util.HadoopInputFile;
import org.apache.parquet.io.InputFile;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.exasol.ExaIterator;
import com.exasol.cloudetl.ScalaConverters;
import com.exasol.cloudetl.bucket.Bucket;
import com.exasol.cloudetl.parquet.ParquetValueConverter;
import com.exasol.cloudetl.source.Source;
import com.exasol.cloudetl.storage.FileFormat;
import com.exasol.cloudetl.storage.StorageProperties;
import com.exasol.cloudetl.transform.DefaultTransformation;
import com.exasol.common.data.Row;
import com.exasol.errorreporting.ExaError;
import com.exasol.parquetio.data.ChunkInterval;
import com.exasol.parquetio.reader.RowParquetChunkReader;
import com.exasol.parquetio.reader.RowParquetReader;

/** Emits file data into Exasol. */
public final class FilesDataEmitter implements Emitter {
    private static final Logger LOGGER = LoggerFactory.getLogger(FilesDataEmitter.class);

    private final StorageProperties properties;
    private final Map<String, List<ChunkInterval>> files;
    private final Bucket bucket;
    private final FileFormat fileFormat;
    private final DefaultTransformation defaultTransformation;

    /** Create an emitter. */
    public FilesDataEmitter(final StorageProperties properties,
            final scala.collection.immutable.Map<String, List<ChunkInterval>> files) {
        this.properties = properties;
        this.files = ScalaConverters.javaMapCopy(files);
        this.bucket = Bucket.create(properties);
        this.fileFormat = properties.getFileFormat();
        this.defaultTransformation = new DefaultTransformation(properties);
    }

    @Override
    public void emit(final ExaIterator context) {
        setDefaultTimezoneToUtcIfEnabled();
        emitFileFormatData(context);
    }

    private void setDefaultTimezoneToUtcIfEnabled() {
        if (this.properties.isEnabled("TIMEZONE_UTC")) {
            TimeZone.setDefault(TimeZone.getTimeZone("UTC"));
        }
    }

    private void emitFileFormatData(final ExaIterator context) {
        if (this.fileFormat == FileFormat.PARQUET || this.fileFormat == FileFormat.DELTA) {
            emitParquetData(context);
        } else {
            emitRegularData(context);
        }
    }

    private void emitRegularData(final ExaIterator context) {
        int totalRowCount = 0;
        for (final String filename : this.files.keySet()) {
            try (Source source = Source.create(this.fileFormat, new Path(filename), this.bucket.getConfiguration(),
                    this.bucket.fileSystem())) {
                int rowCount = 0;
                final scala.collection.Iterator<Row> stream = source.stream();
                while (stream.hasNext()) {
                    final Object[] values = this.defaultTransformation.transform(transformRegularRowValues(stream.next()));
                    emitRow(context, values);
                    rowCount++;
                }
                totalRowCount += rowCount;
                LOGGER.info("Imported file {} with {} rows", filename, rowCount);
            } catch (final Exception exception) {
                throw new IllegalStateException(exception);
            }
        }
        LOGGER.info("Imported {} files with {} rows in total", this.files.size(), totalRowCount);
    }

    private Object[] transformRegularRowValues(final Row row) {
        return ScalaConverters.asJavaList(row.getValues()).toArray(new Object[0]);
    }

    private void emitParquetData(final ExaIterator context) {
        int totalRowCount = 0;
        int totalIntervalCount = 0;
        for (final Map.Entry<String, List<ChunkInterval>> entry : this.files.entrySet()) {
            final String filename = entry.getKey();
            final List<ChunkInterval> intervals = entry.getValue();
            final InputFile inputFile = getInputFile(filename);
            try {
                final ParquetValueConverter converter = new ParquetValueConverter(RowParquetReader.getSchema(inputFile));
                final RowParquetChunkReader source = new RowParquetChunkReader(inputFile, intervals);
                final int[] rowCount = { 0 };
                source.read(row -> {
                    final Object[] values = defaultTransformation.transform(converter.convert(row));
                    emitRow(context, values);
                    rowCount[0]++;
                });
                totalRowCount += rowCount[0];
                totalIntervalCount += intervals.size();
                LOGGER.info("Imported file {} with {} rows and {} intervals", inputFile, rowCount[0], intervals.size());
            } catch (final IOException exception) {
                throw new IllegalStateException(exception);
            }
        }
        LOGGER.info("Imported {} files with {} intervals and {} rows in total", this.files.size(), totalIntervalCount,
                totalRowCount);
    }

    private InputFile getInputFile(final String filename) {
        try {
            return HadoopInputFile.fromPath(new Path(filename), this.bucket.getConfiguration());
        } catch (final IOException exception) {
            throw new IllegalStateException(ExaError.messageBuilder("E-CSE-27")
                    .message("Failed to create an input file from {{FILE}}.", filename)
                    .mitigation("Please make sure that the file is not corrupted.").toString(), exception);
        }
    }

    private static void emitRow(final ExaIterator context, final Object... values) {
        try {
            context.emit(values);
        } catch (final Exception exception) {
            throw new IllegalStateException(exception);
        }
    }
}
